pints/_mcmc/_dual_averaging.py

```diff
@@ -0,0 +1,241 @@
+#
+# Dual Averaging step size and mass matrix adaption method for NUTS and HMC
+# samplers
+#
+# This file is part of PINTS (https://github.com/pints-team/pints/) which is
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+import numpy as np
+
+
+class DualAveragingAdaption:
+    """
+    Implements a Dual Averaging scheme to adapt the step size ``epsilon``, as
+    per [1]_, and estimates the (fully dense) inverse mass matrix using the
+    sample covariance of the accepted parameter, as suggested in [2]_
+
+    The adaption is done using the same windowing method employed by STAN,
+    which is done over three or more windows:
+
+    - initial window: epsilon is adapted using dual averaging
+    - base window: epsilon continues to be adapted using dual averaging, this
+      adaption completes at the end of this window. The inverse mass matrix is
+      adaped at the end of the window by taking the sample covariance of all
+      parameter points in this window.
+    - terminal window: epsilon is adapted using dual averaging, which completes
```

Comment:
I have added that the mass matrix is constant in this terminal window. Have a look to check I haven't spoken out of turn though...

Reply:
yup, this is correct